algo/src/main/java/org/neo4j/graphalgo/impl/walking/NodeWalker.java

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    private long[] doWalk(int startNodeId, int steps, NodeWalker.NextNodeStrategy nextNodeStrategy, Graph graph, TerminationFlag terminationFlag) {
+        long[] nodeIds = new long[steps + 1];
+        int currentNodeId = startNodeId;
+        int previousNodeId = currentNodeId;
+        nodeIds[0] = toOriginalNodeId(graph, currentNodeId);
+        for(int i = 1; i <= steps; i++){
+            int nextNodeId = nextNodeStrategy.getNextNode(currentNodeId, previousNodeId);
+            previousNodeId = currentNodeId;
+            currentNodeId = nextNodeId;
+
+            if (currentNodeId == -1 || !terminationFlag.running()) {
+                // End walk when there is no way out and return empty result
+                return Arrays.copyOf(nodeIds,1);
+            }
+            nodeIds[i] = toOriginalNodeId(graph, currentNodeId);
+        }
+
+        return nodeIds;
+    }
+
+    private long toOriginalNodeId(Graph graph, int currentNodeId) {
+        return currentNodeId == -1 ? -1 : graph.toOriginalNodeId(currentNodeId);
+    }
+
+    /*
+    private void startSingleWalk(AbstractWalkOutput output, int nodeId, int steps, long numberOfElements, long startTime) {
+        try {
+            long[] pathIds = doWalk(nodeId, steps, nextNodeStrategy);
+            long numberOfResults = output.numberOfResults() + 1;
+            // log progress ------
+            if(numberOfResults % 50000 == 2500){ //get a better estimate after 2500 walks
+                long progress = (numberOfResults*1000) / (numberOfElements*10); // dont convert to float but keep precision
+                long remainingTime = estimateRemainingTime(startTime, numberOfResults, numberOfElements);
+                long remainingMinutes = TimeUnit.MINUTES.convert(remainingTime, TimeUnit.NANOSECONDS);
+                long remainingHours = TimeUnit.HOURS.convert(remainingTime, TimeUnit.NANOSECONDS);
+                String progressLog = String.format( "Approximately %d %% of all walks done; estimated remaining time is %d minutes or %d hours", progress, remainingMinutes, remainingHours);
+                log.info(progressLog);
+            }
+            // ------
+
+            output.addResult(pathIds);
+        } catch (Exception e) {
+            log.error(String.format("Error with node %d: %s", nodeId, e.getMessage()));
+        }
+    }
+
+    private long estimateRemainingTime(long startNanoTime, long currentIndex, long size) {
+        long elapsedTime = System.nanoTime() - startNanoTime;
+        double timeForOne = elapsedTime / currentIndex;
+        long remainingElements = size - currentIndex;
+        return remainingElements * (long) timeForOne;
+    }
+    */
+
+    public static class RandomNextNodeStrategy extends NextNodeStrategy {
+
+        public RandomNextNodeStrategy(Graph graph, Degrees degrees) {
+            super(graph, degrees);
+        }
+
+        @Override
+        public int getNextNode(int currentNodeId, int previousNodeId) {
```

Comment:
If I'm reading correctly both of the provided strategies select one of a node's relationships to follow. Should we have a strategy where part of the random walk could involve jumping to a completely different node and continuing the walk from there?

Reply:
I was thinking about that, the main drawback I see is that it would kinda break the "path" in the graph but if we use virtual rels anyway that might be fine.
Not sure if we should somehow indicate in the results where such a jump occurred?